### ISSUE TITLE
Where to modify the code of the model?

### ISSUE THREAD

**Issue by othsueh** [ZebangCheng/Emotion-LLaMA#27]
Hello author, I am currently working on using IEMOCAP dataset with multi-label approach on your architecture, with audio, video and text as input. But I ran in some problems with your code, here are the problems:

1. I have seen documents from [MiniGPT-4](https://github.com/Vision-CAIR/MiniGPT-4), it seems like you use finetune of MiniGPT-4 approach and follow its framework of setting yaml in `minigpt4/config/datasets`. But I am wondering that is the image_path still needed? Since there is an error occur from this.
https://github.com/ZebangCheng/Emotion-LLaMA/blob/8558bd98dd05654467972091a80b3c12ca199bc8/minigpt4/configs/datasets/firstface/featureface.yaml#L5-L9
2. Since I use multi-label approach, so the way to calculate the loss is different. Where is the part you calculate error between prediction and label (from `MERR_coarse_grained.txt`) to backpropagate to linear layer? I have tracked it for a long time but still can't find it.

Be grateful for your upcoming reply, thank you!

**Comment by ZebangCheng** [comment]
Hi, 

Thank you for your questions and for using our architecture with the IEMOCAP dataset! Below are the responses to your concerns:

1. **Regarding `image_path`:**  
   Due to privacy usage agreements associated with the dataset, we cannot release the extracted face images publicly. Therefore, we have updated the relevant code to read directly from the source videos instead of using pre-extracted images. The disclosure of the old code was an oversight during our merge process. Thank you for pointing this out, and we have now updated the repository accordingly. The relevant code has been updated as follows:

https://github.com/ZebangCheng/Emotion-LLaMA/blob/190734dd5bca5df870897795a75a522d845381ab/minigpt4/configs/datasets/firstface/featureface.yaml#L5-L9

https://github.com/ZebangCheng/Emotion-LLaMA/blob/190734dd5bca5df870897795a75a522d845381ab/minigpt4/datasets/datasets/first_face.py#L98-L117

2. **About calculating loss for multi-label tasks:**  
   There might be some misunderstandings that I’d like to clarify:  
   
   2.1 Initially, we tried classifying the hidden features from the large model using a linear layer, but this approach showed suboptimal performance. As a result, we shifted to having the large model directly output emotion labels as text (e.g., happy, sad, fear, ...) instead of traditional discrete numerical categories (e.g., 0, 1, 2, ...). For this, we used the standard Cross-Entropy Loss (CELoss) commonly employed for large language models, calculating the accuracy of predicted text-based emotion labels.  

   2.2 For multi-label tasks, we have performed experiments on tasks such as MER-OV (open-vocabulary emotion recognition). The main difference is that in MER-OV, there is no restriction on emotion categories. For the IEMOCAP dataset with a multi-label approach, you can constrain the model to output specific emotion categories and craft the prompts to include multiple emotion labels. Here is an example of how we structure the prompts:  

   ``` 
   # Single-label example  
   Prompt: "Please determine which emotion label in the video represents: happy, sad, neutral, angry, worried, surprise."  
   Instruction: "surprise"  

   # Multi-label example  
   Prompt: "Please determine which emotion label in the video represents: happy, sad, neutral, angry, worried, surprise. You need to provide all relevant emotion labels displayed in the video."  
   Instruction: "surprise, happy"  
   ```
   The Multi-label example mentioned above has not yet been implemented in Emotion-LLaMA. It is a simple concept of mine that requires constructing a relevant instruction dataset for instruction-tuning.

Feel free to reach out if you have any further questions or need more clarification. I hope this helps!

**Comment by othsueh** [comment]
Thank you for your answer, it's really helpful!
Recently I am working on transform your `first_face.py` file to fit IEMOCAP dataset, and I accidentally find out some bugs, I have made the PR, please check it. 
1. I want to ask where is the part that you read LLM's output and compare it with label? Since with multi-lable approach, I need to transform the output e.g. "surprise, happy" to an array of emotion distribution.
2. I don't understand the logic of using text_preprocessor, what's it for and why it doesn't use in infer task. And also what's the difference between caption and emotion(emo2idx)? Since caption in originally given data is also emotion,  what's the function of them in training?
https://github.com/ZebangCheng/Emotion-LLaMA/blob/cad86fd7ed2948a848d5d71ba598b3e1c5d652fb/minigpt4/datasets/datasets/first_face.py#L131-L158 
Be grateful for your upcoming reply, thank you!

**Comment by ZebangCheng** [comment]
1. **LLMs Loss Computation**  
   The loss computation for MLLMs (Multimodal Large Language Models) is described in detail in the following code:  
https://github.com/ZebangCheng/Emotion-LLaMA/blob/cad86fd7ed2948a848d5d71ba598b3e1c5d652fb/minigpt4/models/modeling_llama.py#L87-L102 

   If you want to transform the output (e.g., "surprise, happy") into an array of emotion distributions, you should consider modifying the dataset's `answer` field accordingly and retraining the model. Such as `"answer": "surprise, happy"`. This will guide the model to output the desired format during inference. Please refer to this section for dataset adjustment:
  https://github.com/ZebangCheng/Emotion-LLaMA/blob/cad86fd7ed2948a848d5d71ba598b3e1c5d652fb/minigpt4/datasets/datasets/first_face.py#L164-L172

2. **Functionality of `text_preprocessor`**  
   You can observe the effects of using `text_preprocessor` by comparing the input and output strings as shown below:  
   ```python
   print(f"instruction_input: {instruction_input}")  
   # instruction_input: <video><VideoHere></video> <feature><FeatureHere></feature> [emotion] Please determine which emotion label in the video represents: happy, sad, neutral, angry, worried, surprise.  
   print(f"self.text_processor(instruction_input): {self.text_processor(instruction_input)}")  
   # self.text_processor(instruction_input): <video><videohere></video> <feature><featurehere></feature> [emotion] please determine which emotion label in the video represents happy, sad, neutral, angry, worried, surprise  
   ```  
   The `text_preprocessor` standardizes text by converting all characters to lowercase and removing punctuation like periods (`.`).  

3. **Difference Between "emotion" and "caption"**  
   - The `emotion` field in the dataset corresponds to discrete numerical labels (e.g., 0, 1, 2, ...) for emotion categories. These were part of earlier implementations, and I apologize for any confusion caused.  
   - The `caption` field contains the content the model is trained to predict. For instance:  
     - In **emotion recognition** tasks (`[emotion]`), `caption` includes textual emotion labels (e.g., "happy").  
     - In **emotion reasoning** tasks (`[reason]`), `caption` may include tone descriptions, facial expression details, etc.  
   
These distinctions can be explored by printing or debugging the dataset during preprocessing.

**Comment by 00dbgpdnjs** [comment]
@othsueh Did you successfully apply it to the IEMOCAP dataset? I'm also interested in applying it to a different dataset, and I'm curious how you approached it.
